File: apps/macos/LexiconBar/Tests/LexiconBarKitTests/PartialWriteTests.swift
import XCTest
@testable import LexiconBarKit

/// The arithmetic behind "the correction only half went in".
///
/// This is the one failure in the app that can destroy text the user wrote.
/// The keystroke write posts a run of `CGEvent`s, so a correction over twenty
/// units is several of them; a chunk that could not be posted used to be
/// skipped in silence, and the engine then compared the field against the value
/// the plan was made from, found it different and told the user the write had
/// failed. The field held half a replacement, the original span was gone and
/// nothing had been recorded that could put it back.
///
/// What a short write left behind is computed here. Every case below asserts
/// the same two things: the field text the engine will check its read against,
/// and that undoing the entry restores exactly what the user had. The second is
/// the one that matters.
final class PartialWriteTests: XCTestCase {
    private let key = "field-1"
    private let before = "I use ashler daily"

    /// "I use ashler daily" with "ashler" becoming "Ashlr.AI".
    private func midFieldPlan() -> RewritePlan.Plan {
        plan(before: before, range: NSRange(location: 6, length: 6), replacement: "Ashlr.AI")
    }

    private func plan(before: String, range: NSRange, replacement: String) -> RewritePlan.Plan {
        RewritePlan.Plan(range: range,
                         previousText: TextDiff.substring(before, utf16: range)!,
                         newText: replacement,
                         splicedFullText: TextDiff.splice(before, utf16: range, with: replacement)!)
    }

    func testHalfATypedReplacementIsDescribedAndUndoable() throws {
        // "Ashl" of "Ashlr.AI" landed; the selected "ashler" is already gone.
        let partial = try XCTUnwrap(PartialWrite.overSelection(plan: midFieldPlan(), before: before, unitsTyped: 4))

        XCTAssertEqual(partial.fieldText, "I use Ashl daily")
        XCTAssertEqual(partial.writtenText, "Ashl")
        XCTAssertEqual(partial.originalText, "ashler")
        XCTAssertEqual(partial.undone, before)
    }

    func testTheEntryPutsTheUsersOwnTextBack() throws {
        let partial = try XCTUnwrap(PartialWrite.overSelection(plan: midFieldPlan(), before: before, unitsTyped: 4))

        let entry = partial.entry(fieldKey: key)
        XCTAssertEqual(entry.fieldKey, key)
        XCTAssertEqual(entry.fieldTextAfter, partial.fieldText)
        XCTAssertEqual(entry.fieldTextBefore, before)

        // And the ledger will actually offer it: the field holds exactly what
        // the entry says it does.
        var ledger = UndoLedger()
        ledger.record(entry)
        XCTAssertTrue(ledger.canUndo(fieldKey: key, currentText: partial.fieldText))
    }

    func testOneUnitIsStillAPartialWrite() throws {
        let partial = try XCTUnwrap(PartialWrite.overSelection(plan: midFieldPlan(), before: before, unitsTyped: 1))

        XCTAssertEqual(partial.fieldText, "I use A daily")
        XCTAssertEqual(partial.undone, before)
    }

    func testNothingTypedIsNotAPartialWrite() {
        // A selection that was never typed over still holds the user's text.
        XCTAssertNil(PartialWrite.overSelection(plan: midFieldPlan(), before: before, unitsTyped: 0))
        XCTAssertNil(PartialWrite.overSelection(plan: midFieldPlan(), before: before, unitsTyped: -3))
    }

    func testAFullyTypedReplacementIsNotAPartialWrite() {
        let plan = midFieldPlan()
        let units = plan.newText.utf16.count
        XCTAssertNil(PartialWrite.overSelection(plan: plan, before: before, unitsTyped: units))
        XCTAssertNil(PartialWrite.overSelection(plan: plan, before: before, unitsTyped: units + 5))
    }

    func testAFieldThatNoLongerHoldsThePlansSpanIsRefused() {
        // The premise is already false, so there is no state to describe and
        // the engine must leave the field alone rather than guess at one.
        XCTAssertNil(PartialWrite.overSelection(plan: midFieldPlan(), before: "I use something else", unitsTyped: 4))
        XCTAssertNil(PartialWrite.overSelection(plan: midFieldPlan(), before: "short", unitsTyped: 4))
    }

    func testEverySelectionProgressRoundTripsBackToTheOriginal() throws {
        let plan = midFieldPlan()

        for typed in 1..<plan.newText.utf16.count {
            let partial = try XCTUnwrap(PartialWrite.overSelection(plan: plan, before: before, unitsTyped: typed),
                                        "no state described for \(typed) units")
            XCTAssertEqual(partial.undone, before, "\(typed) units")
            XCTAssertEqual(partial.entry(fieldKey: key).fieldTextBefore, before, "\(typed) units")
        }
    }

    /// The write arriving in full after the caller gave up on it. Without a
    /// description of that state the correction would sit in the field with no
    /// ledger entry behind it.
    func testTheWholeCorrectionIsDescribedForALateArrival() {
        let plan = midFieldPlan()
        let whole = PartialWrite.whole(plan)

        XCTAssertEqual(whole.fieldText, plan.splicedFullText)
        XCTAssertEqual(whole.writtenText, plan.newText)
        XCTAssertEqual(whole.undone, before)
        XCTAssertEqual(whole.entry(fieldKey: key).fieldTextBefore, before)
    }

    /// A replacement wider than one UTF-16 unit per character. The undo splices
    /// back over what was written, which is counted in units and not in
    /// characters, so anything that measures it the other way puts the user's
    /// text back in the wrong place.
    func testAstralCharactersRoundTripByUnitsNotCharacters() throws {
        let before = "ship it soon"
        let plan = plan(before: before, range: NSRange(location: 5, length: 2), replacement: "\u{1F680}\u{1F680}ok")

        for typed in stride(from: 2, to: plan.newText.utf16.count, by: 2) {
            let partial = try XCTUnwrap(PartialWrite.overSelection(plan: plan, before: before, unitsTyped: typed))
            XCTAssertEqual(partial.undone, before, "\(typed) units")
            XCTAssertEqual(partial.entry(fieldKey: "f").fieldTextBefore, before, "\(typed) units")
        }
    }
}

File: apps/windows/tests/LexiconBar.Core.Tests/PartialWriteTests.cs
using Xunit;

namespace LexiconBar.Tests;

/// <summary>
/// The arithmetic behind "the correction only half went in".
///
/// This is the one failure in the app that can destroy text the user wrote.
/// The keystroke write used to be a loop of <c>SendInput</c> calls, so a
/// correction over 200 characters was several of them; when a later call was
/// refused the earlier ones had already landed, the engine compared the field
/// against the value the plan was made from, found it different and told the
/// user the write had failed. The field held half a replacement, the original
/// span was gone and nothing had been recorded that could put it back.
///
/// The write is one call now, and what a short call left behind is computed
/// here. Every case below asserts the same two things: the field text the
/// engine will check its read against, and that undoing the entry restores
/// exactly what the user had. The second is the one that matters.
/// </summary>
public class PartialWriteTests
{
    private const string Key = "field-1";

    /// <summary>"I use ashler daily" with "ashler" becoming "Ashlr.AI".</summary>
    private static RewritePlan MidFieldPlan() => Plan(
        before: "I use ashler daily",
        span: new TextSpan(6, 6),
        replacement: "Ashlr.AI");

    /// <summary>The same correction sitting at the very end of the field.</summary>
    private static RewritePlan TailPlan() => Plan(
        before: "I use ashler",
        span: new TextSpan(6, 6),
        replacement: "Ashlr.AI");

    private static RewritePlan Plan(string before, TextSpan span, string replacement)
    {
        string previous = TextDiff.Substring(before, span)!;
        string spliced = TextDiff.Splice(before, span, replacement)!;
        return new RewritePlan(span, previous, replacement, spliced);
    }

    // ------------------------------------------------------- select and type

    [Fact]
    public void HalfATypedReplacementIsDescribedAndUndoable()
    {
        const string before = "I use ashler daily";
        RewritePlan plan = MidFieldPlan();

        // "Ashl" of "Ashlr.AI" landed; the selected "ashler" is already gone.
        PartialWrite partial = PartialWrite.OverSelection(plan, before, 4)!;

        Assert.Equal("I use Ashl daily", partial.FieldText);
        Assert.Equal("Ashl", partial.WrittenText);
        Assert.Equal("ashler", partial.OriginalText);
        Assert.Equal(before, partial.Undone);
    }

    [Fact]
    public void TheEntryPutsTheUsersOwnTextBack()
    {
        const string before = "I use ashler daily";
        PartialWrite partial = PartialWrite.OverSelection(MidFieldPlan(), before, 4)!;

        UndoLedger.Entry entry = partial.Entry(Key);
        Assert.Equal(Key, entry.FieldKey);
        Assert.Equal(partial.FieldText, entry.FieldTextAfter);
        Assert.Equal(before, entry.FieldTextBefore);

        // And the ledger will actually offer it: the field holds exactly what
        // the entry says it does.
        UndoLedger ledger = new();
        ledger.Record(entry);
        Assert.True(ledger.CanUndo(Key, partial.FieldText));
    }

    [Fact]
    public void OneUnitIsStillAPartialWrite()
    {
        PartialWrite partial = PartialWrite.OverSelection(MidFieldPlan(), "I use ashler daily", 1)!;

        Assert.Equal("I use A daily", partial.FieldText);
        Assert.Equal("I use ashler daily", partial.Undone);
    }

    [Fact]
    public void NothingTypedIsNotAPartialWrite()
    {
        // A selection that was never typed over still holds the user's text.
        Assert.Null(PartialWrite.OverSelection(MidFieldPlan(), "I use ashler daily", 0));
        Assert.Null(PartialWrite.OverSelection(MidFieldPlan(), "I use ashler daily", -3));
    }

    [Fact]
    public void AFullyTypedReplacementIsNotAPartialWrite()
    {
        RewritePlan plan = MidFieldPlan();
        Assert.Null(PartialWrite.OverSelection(plan, "I use ashler daily", plan.NewText.Length));
        Assert.Null(PartialWrite.OverSelection(plan, "I use ashler daily", plan.NewText.Length + 5));
    }

    [Fact]
    public void AFieldThatNoLongerHoldsThePlansSpanIsRefused()
    {
        // The premise is already false, so there is no state to describe and
        // the engine must leave the field alone rather than guess at one.
        Assert.Null(PartialWrite.OverSelection(MidFieldPlan(), "I use something else", 4));
        Assert.Null(PartialWrite.OverSelection(MidFieldPlan(), "short", 4));
    }

    // ---------------------------------------------------- backspace and type

    [Fact]
    public void BackspacesThatLandedWithoutTheirRetypeAreRecorded()
    {
        const string before = "I use ashler";
        RewritePlan plan = TailPlan();

        // Six key events: every backspace, none of the replacement. This is the
        // state the old two-call form could leave behind unconditionally.
        PartialWrite partial = PartialWrite.OverTail(plan, before, 6)!;

        Assert.Equal("I use ", partial.FieldText);
        Assert.Equal(string.Empty, partial.WrittenText);
        Assert.Equal("ashler", partial.OriginalText);
        Assert.Equal(before, partial.Undone);
        Assert.Equal(before, partial.Entry(Key).FieldTextBefore);
    }

    [Fact]
    public void SomeBackspacesAndSomeRetypeAreRecorded()
    {
        const string before = "I use ashler";
        RewritePlan plan = TailPlan();

        // Nine events: six backspaces, then "Ash".
        PartialWrite partial = PartialWrite.OverTail(plan, before, 9)!;

        Assert.Equal("I use Ash", partial.FieldText);
        Assert.Equal("Ash", partial.WrittenText);
        Assert.Equal("ashler", partial.OriginalText);
        Assert.Equal(before, partial.Undone);
    }

    [Fact]
    public void PartOfTheBackspacesIsRecorded()
    {
        const string before = "I use ashler";

        // Two events: two backspaces, nothing typed. Only "er" is gone.
        PartialWrite partial = PartialWrite.OverTail(TailPlan(), before, 2)!;

        Assert.Equal("I use ashl", partial.FieldText);
        Assert.Equal("er", partial.OriginalText);
        Assert.Equal(string.Empty, partial.WrittenText);
        Assert.Equal(before, partial.Undone);
    }

    [Fact]
    public void AllTheEventsOrNoneIsNotAPartialWrite()
    {
        RewritePlan plan = TailPlan();
        int events = plan.PreviousText.Length + plan.NewText.Length;

        Assert.Null(PartialWrite.OverTail(plan, "I use ashler", 0));
        Assert.Null(PartialWrite.OverTail(plan, "I use ashler", events));
        Assert.Null(PartialWrite.OverTail(plan, "I use ashler", events + 1));
    }

    [Fact]
    public void TheTailPathRefusesAPlanThatIsNotTheTail()
    {
        // The backspace strategy only runs on the tail of the field. Anywhere
        // else the backspaces do not eat the plan's span, so the arithmetic
        // here would be a fiction.
        Assert.Null(PartialWrite.OverTail(MidFieldPlan(), "I use ashler daily", 6));
    }

    [Fact]
    public void EveryTailProgressRoundTripsBackToTheOriginal()
    {
        const string before = "I use ashler";
        RewritePlan plan = TailPlan();
        int events = plan.PreviousText.Length + plan.NewText.Length;

        for (int delivered = 1; delivered < events; delivered++)
        {
            PartialWrite partial = PartialWrite.OverTail(plan, before, delivered)!;
            Assert.Equal(before, partial.Undone);
            Assert.Equal(before, partial.Entry(Key).FieldTextBefore);
        }
    }

    [Fact]
    public void EverySelectionProgressRoundTripsBackToTheOriginal()
    {
        const string before = "I use ashler daily";
        RewritePlan plan = MidFieldPlan();

        for (int typed = 1; typed < plan.NewText.Length; typed++)
        {
            PartialWrite partial = PartialWrite.OverSelection(plan, before, typed)!;
            Assert.Equal(before, partial.Undone);
            Assert.Equal(before, partial.Entry(Key).FieldTextBefore);
        }
    }
}
